File: docs/archive/agent-orchestration/phase0-proposal/canonical/canonical_json.py
"""Proposal-local implementation of tilesim.bridge.canonical_json.v1."""

from __future__ import annotations

import hashlib
import json


def canonical_json_text(value: object) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        raise TypeError("canonical JSON forbids binary floating-point values")
    if isinstance(value, list):
        return "[" + ",".join(canonical_json_text(item) for item in value) + "]"
    if isinstance(value, dict):
        if not all(isinstance(key, str) for key in value):
            raise TypeError("canonical JSON object keys must be strings")
        return "{" + ",".join(
            canonical_json_text(key) + ":" + canonical_json_text(value[key])
            for key in sorted(value, key=lambda item: tuple(ord(character) for character in item))
        ) + "}"
    raise TypeError(f"unsupported canonical JSON type: {type(value).__name__}")


def canonical_sha256(value: object) -> str:
    return "sha256:" + hashlib.sha256(canonical_json_text(value).encode("utf-8")).hexdigest()


def digest_material(document: dict) -> dict:
    fields = document["digest_material_fields"]
    if "canonical_digest" in fields or len(fields) != len(set(fields)):
        raise ValueError("digest material fields must be unique and exclude canonical_digest")
    missing = [field for field in fields if field not in document]
    if missing:
        raise ValueError(f"digest material fields are missing: {missing}")
    return {field: document[field] for field in fields}

File: docs/archive/agent-orchestration/phase0-proposal/canonical/canonical_json.ts
import { createHash } from "node:crypto";

function compareCodePoints(left: string, right: string): number {
  const a = Array.from(left, (character) => character.codePointAt(0) as number);
  const b = Array.from(right, (character) => character.codePointAt(0) as number);
  for (let index = 0; index < Math.min(a.length, b.length); index += 1) {
    if (a[index] !== b[index]) return a[index] - b[index];
  }
  return a.length - b.length;
}

export function canonicalJsonText(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "string") return JSON.stringify(value);
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "bigint") return value.toString(10);
  if (typeof value === "number") {
    if (!Number.isFinite(value) || !Number.isSafeInteger(value)) {
      throw new TypeError("canonical JSON requires a finite lossless integer or decimal string");
    }
    return Object.is(value, -0) ? "0" : JSON.stringify(value);
  }
  if (Array.isArray(value)) return `[${value.map(canonicalJsonText).join(",")}]`;
  if (typeof value === "object") {
    const record = value as Record<string, unknown>;
    return `{${Object.keys(record)
      .sort(compareCodePoints)
      .map((key) => `${JSON.stringify(key)}:${canonicalJsonText(record[key])}`)
      .join(",")}}`;
  }
  throw new TypeError(`unsupported canonical JSON type: ${typeof value}`);
}

export function canonicalSha256(value: unknown): string {
  return `sha256:${createHash("sha256").update(canonicalJsonText(value), "utf8").digest("hex")}`;
}

export function digestMaterial(document: Record<string, unknown>): Record<string, unknown> {
  const fields = document.digest_material_fields;
  if (!Array.isArray(fields) || !fields.every((field) => typeof field === "string")) {
    throw new TypeError("digest_material_fields must be a string array");
  }
  if (fields.includes("canonical_digest") || new Set(fields).size !== fields.length) {
    throw new TypeError("digest material fields must be unique and exclude canonical_digest");
  }
  return Object.fromEntries(
    fields.map((field) => {
      if (!(field in document)) throw new TypeError(`missing digest material field: ${field}`);
      return [field, document[field]];
    }),
  );
}
